agenthub/codeact_agent/codeact_agent.py

```diff
@@ -194,27 +215,148 @@ def step(self, state: State) -> Action:
         if latest_user_message and latest_user_message.strip() == '/exit':
             return AgentFinishAction()
 
-        # prepare what we want to send to the LLM
-        messages: list[dict[str, str]] = self._get_messages(state)
-
-        response = self.llm.completion(
-            messages=messages,
-            stop=[
-                '</execute_ipython>',
-                '</execute_bash>',
-                '</execute_browse>',
-            ],
-            temperature=0.0,
-        )
+        response = None
+        # give it multiple chances to get a response
+        # if it fails, we'll try to condense memory
+        attempt = 0
+        while not response and attempt < self.attempts_to_condense:
+            # prepare what we want to send to the LLM
+            messages: list[dict[str, str]] = self._get_messages(state)
+            print('No of tokens, ' + str(self.llm.get_token_count(messages)) + '\n')
+            try:
+                if self.llm.is_over_token_limit(messages):
+                    raise TokenLimitExceededError()
+                response = self.llm.completion(
```

Comment:
I'm thinking if it will be cleaner that we move these logic into the `LLM` class?
I'm thinking we could add an argument to `llm.completion` say `condense=True` which will trigger the `llm.completion` logic. This way, we can automatically supports a lot of different agents without writing duplicated code.